### PR TITLE
Start magnetissimo in a Docker container

### PR DESCRIPTION
**Pull request by luciano-fiandesio**
Hi,
I have created a dockerfile and a docker-compose file to dockerize magetissimo.
The docker-compose file will start 2 containers, one for Postgres and one for magnetissimo, and one persistent volume, for the data.

`docker-compose up -d`

http://localhost:4000

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+RUN mix local.rebar --force
+
+# Install the Phoenix framework itself
+RUN mix archive.install --force https://github.com/phoenixframework/archives/raw/master/phoenix_new.ez
+
+# Set /app as workdir
+WORKDIR /app
+
+# disable ssl verify for git
+RUN git config --system http.sslVerify false
```
Comment (evaryont) [inline docker/Dockerfile]: why are you disabling SSL verification? That really shouldn't be necessary.

Comment (luciano-fiandesio) [inline docker/Dockerfile]: Yep, not needed, I have pushed a change to Dockerfile, where the ca-certificates are installed in the base image. No more ssl ignoring

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
+FROM mrrooijen/phoenix:latest
```
Comment (Nikkau) [inline docker/Dockerfile]: This base image contains a lot of things Magnetissimo doesn't need, like Node and a bunch of Erlang stuff.
I think starting fresh from alpine and installing only required stuff is better.

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+# Install certificates
+RUN apk add ca-certificates
+
+RUN mix local.hex --force
```
Comment (Nikkau) [inline docker/Dockerfile]: Only needed if the parent image is old, like in my previous comment, I think we should directly build all stuff in this image.

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+# Install rebar
+RUN mix local.rebar --force
```
Comment (Nikkau) [inline docker/Dockerfile]: Same ad `local.hex`.

File: docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+# Install the Phoenix framework itself
+RUN mix archive.install --force https://github.com/phoenixframework/archives/raw/master/phoenix_new.ez
```
Comment (Nikkau) [inline docker/Dockerfile]: The `mix deps.get` should do this work.

BTW, it's a bad practice, `mix.exs` tell the version the app need, it's ok to rely on an old one.

File: docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+# Set /app as workdir
+WORKDIR /app
+
+# fetch magnetissimo
+RUN git clone https://github.com/sergiotapia/magnetissimo.git /app
```
Comment (Nikkau) [inline docker/Dockerfile]: You should `COPY` to import current filesystem, we should be able to build the current branch/MR.

File: docker/prod.secret.exs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+use Mix.Config
+
+config :magnetissimo, Magnetissimo.Repo,
+  adapter: Ecto.Adapters.Postgres,
+  username: "torrent",
+  password: "postgres-secret",
+  database: "magnetissimo_dev",
```
Comment (Nikkau) [inline docker/prod.secret.exs]: The Docker build shouldn't rely on a special config file.

Instead the prod config file should read environment variables:
- The `docker-compose.yml` can pass good values.
- The config file can be use by other methods of deployement, like Heroku.

Comment (luciano-fiandesio) [inline docker/prod.secret.exs]: How would you do that?

Comment (Nikkau) [inline docker/prod.secret.exs]: You have some examples in https://github.com/sergiotapia/magnetissimo/blob/master/config/prod.exs
`http: [port: {:system, "PORT"}]`

**Comment by sergiotapia** [comment]
@luciano-fiandesio Please let me know when you feel this PR is at a good point so I can merge it in. I'm not experienced with Docker at all. 😉

**Comment by luciano-fiandesio** [comment]
Hi, 

Yes,  I did the same.  I'm using compose to start 2 containers and a volume,  for postgres data. 

ddmng <notifications@github.com> wrote:
>@luciano-fiandesio @Nikkau
>I've worked on my fork yesterday to dockerize magnetissimo but I've
>seen only today this PR. If it can be of any help, I've successfully
>set up a compose that starts postgres and magnetissimo in dev
>environment. prod and test should be straightforward to achieve:
>https://github.com/ddmng/magnetissimo
>
>—
>You are receiving this because you were mentioned.
>Reply to this email directly, view it on GitHub, or mute the thread.

I prefer to use encrypted email. My public key fingerprint is 25C7 8A5F E43C 09BB AE63  616F 3023 16AA E102 CD74. Learn how to encrypt your email with the Email Self Defense guide.

**Comment by christian-fei** [comment]
any news on this?

**Comment by sergiotapia** [comment]
Yeah guys, as mentioned before: Let me know when it's good to merge. 😄  Thank you!

**Comment by luciano-fiandesio** [comment]
Hey guys, sorry, I'm a bit snowed under at this moment (moving to a different country with wife and kids...). For me the Docker config is good enough as it is now, if someone else wants to change the base image or some other minor details, go ahead.

**Comment by evaryont** [comment]
@luciano-fiandesio I have some ideas to improve this. If you're willing, I can submit the changes to your PR.

**Comment by luciano-fiandesio** [comment]
@evaryont sure, go ahead

**Comment by knaperek** [comment]
Anybody else getting this error when trying to build the image?
```
Cloning into '/app'...
fatal: unable to access 'https://github.com/sergiotapia/magnetissimo.git/': error setting certificate verify locations:
  CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
ERROR: Service 'web' failed to build: The command '/bin/sh -c git clone https://github.com/sergiotapia/magnetissimo.git /app' returned a non-zero code: 128
```

**Comment by CrazyGentleMan** [comment]
@knaperek I got the same error.

I fixed it by add `RUN git config --global http.sslverify "false"` before cloning.

**Comment by Nikkau** [comment]
Hi,
I still think my way is better, I don't want clone `master` when I work in a feature branch.

File: docker/Dockerfile (hunk 2 above)
Comment (zweicoder) [inline docker/Dockerfile]: Isn't Erlang required by Elixir and Node required by Phoenix?

**Comment by zweicoder** [comment]
If anyone is interested I worked on a fork [here](https://github.com/zweicoder/magnetissimo/commits/docker) taking into account some of the responses here

**Comment by pahakalle** [comment]
@zweicoder Why not merge it here ?

**Comment by firedomain** [comment]
Is there any update on official docker support?

**Comment by pahakalle** [comment]
@sergiotapia Any new developments towards docker support?

**Comment by firedomain** [comment]
Any news? have been trying to install from 3rd party docker repos & have had issues getting it running... really hoping for some official support.

**Comment by tchoutri** [comment]
@firedomain Hi, I'll work on an official Dockerfile today, thanks for the reminder.

xoxo

**Comment by alpharde** [comment]
Hi, I think I got a docker image going but with a few caveats.
Please bear with me as I don't know anything about erlang:
- If I understood corretly _mix ecto.migrate_ compiles some stuff and create tables, can I get those two to work sepparately? Compilation should happen while the image is building and table creation/update should happen during runtime.
- The application itself is running and downloading but I get no webinterface and the program doesn't bind to port 4000, I might need to test this sepparately on LXC or something.

**Comment by skwerlman** [comment]
`mix` is part of the elixir tooling, not part of erlang

`mix ecto.migrate` just runs migrations. if mix determines that code needs to be compiled, it runs `mix compile`. if you want it to happen earlier, just call `mix compile` as part of the image creation.

also, `mix ecto.create` is what creates the tables; `mix ecto.migrate` is probably calling it implicitly. it might be better to make it explicit instead.

not sure why you aren't getting a webui; i don't have too much experience with containers.

**Comment by skwerlman** [comment]
are you setting `PORT=4000` when compiling & running? the prod config expects that in the environment, while dev always uses 4000

**Comment by alpharde** [comment]
Yes, PORT is set up to 4000.
Looking foward to it.

**Pull request closed without merging**